text/0000-epochs.md

```diff
@@ -0,0 +1,543 @@
+
+- First, introduce and stabilize `dyn Trait`.
+- Deprecate bare `Trait` syntax in favor of `dyn Trait`.
+- In a preview epoch, make it an error to use bare `Trait` syntax.
+- Ship the new epoch, and wait until bare `Trait` syntax is obscure.
+- Re-introduce bare `Trait` syntax, stabilize it, and deprecate `impl Trait` in
+  favor of it.
+
+Of course, this RFC isn't suggesting that such a course of action is a *good*
+one, just that it is *possible* to do without breakage.
+
+#### Example: type inference changes
+
+There are a number of details about type inference that seem suboptimal:
+
+- Currently multi-parameter traits like `AsRef<T>` will infer the value of one
+  parameter on the basis of the other. We would at least like an opt-out, but
+  employing it for `AsRef` is backwards-incompatible.
+- Coercions don’t always trigger when we wish they would, but altering the rules
+  may cause other programs to stop compiling.
+- In trait selection, where-clauses take precedence over impls; changing this is backwards-incompatible.
+
+We may or may not be able to change these details on the existing epoch. With
+enough effort, we could probably deprecate cases where type inference rules
+might change and request explicit type annotations, and then—in the new
+epoch—tweak those rules.
+
+### What epochs can't do
+
+There are also changes that epochs don't help with, due to the constraints we
+impose. These limitations are extremely important for keeping the compiler
+maintainable, the language understandable, and the ecosystem compatible.
+
+#### Example: changes to coherence rules
+
+Trait coherence rules, like the "orphan" rule, provide a kind of protocol about
+which crates can provide which `impl`s. It's not possible to change protocol
+incompatibly, because existing code will assume the current protocol and provide
+impls accordingly, and there's no way to work around that fact via deprecation.
+
+More generally, this means that epochs can only be used to make changes to the
+language that are applicable *crate-locally*; they cannot impose new
+requirements or semantics on external crates, since we want to retain
+compatibility with the existing ecosystem.
+
+#### Example: `Error` trait downcasting
+
+See [rust-lang/rust#35943](https://github.com/mozilla/rust/issues/35943). Due to
+a silly oversight, you can’t currently downcast the “cause” of an error to
+introspect what it is. We can’t make the trait have stricter requirements; it
+would break existing impls. And there's no way to do so only in a newer epoch,
+because we must be compatible with the older one, meaning that we cannot rely on
+downcasting.
+
+This is essentially another example of a non-crate-local change.
+
+## The full mechanics
+
+We'll wrap up with the full details of the mechanisms at play.
+
+- `rustc` will take a new flag, `--epoch`, which can specify the epoch to
+  use. This flag will default to the current epoch.
```

Comment:
I read through the entire RFC but missed this detail. I saw that Cargo defaults to `epoch = "2015"` and assumed that rustc would too.

I think this should be reconsidered. Rust-aware build systems other than Cargo can also decide to enable the latest epoch as part of their template for new projects.